### ISSUE TITLE
TcpReassembly (question)

### ISSUE THREAD

**Issue by gx740** [seladb/PcapPlusPlus#248]
I have found that the methods `closeConnectionInternal` and `closeAllConnections` have different behavior in relation to `m_ConnectionInfo` vector. The `closeAllConnections` clears this vector while closeConnectionInternal does not remove the data from it.

Should `closeConnectionInternal` remove data from vector or not?

**Comment by gx740** [comment]
I think this is a bug.

**Comment by seladb** [comment]
Yes I think you're right, good job on catching it!

Since I wrote this code a long back, I'd suggest testing the change carefully as I'm not sure if I had a good reason for not removing the `ConnectionData` from `m_ConnectionInfo`

**Comment by gx740** [comment]
I have looked into the code and found that there is one reason not to remove connection info: if the client's code wants to use this info after session removing. There is the method getConnectionInformation to get a const reference to m_ConnectInfo at any time. But I don't see the reasonable use case for such idea. We just get a memory leakage.

**Comment by gx740** [comment]
It's a little about getConnectionInformation method. It seems to me that this method is quite useless. Why? It returns an unsorted vector and we have no way of quick access by flowKey.

**Comment by seladb** [comment]
Thanks for checking, I think that was probably the reason I decided not to remove the connection info. I do agree that in large scale this will cause a memory leak, but maybe there are use cases where you need to query this data. 

I also agree that for a large number of connection using `getConnectionInformation()` will be quite useless. Maybe we should consider replacing the vector with a map?

To summarize, maybe we can add the following features:

- Replace the connection info vector with a map (which its key is flow key)
- Add a flag indicating whether to remove the connection data after a connection is closed
- Add a method for cleaning connection data upon user request

**Comment by gx740** [comment]
Yes we can add these features.
Should the flag be a property of class or a parameter of function?

I think additional optimizations can be done (maybe in another PR):
1. Removing m_ClosedConnectionList at all
2. m_ConnectionList will store a pointer to data or null. Null means the connection is closed.

**Comment by seladb** [comment]
As per my answer in #250 we cannot simply remove `m_ClosedConnectionList`, but we can create a mechanism to clean it a short time after a connection is closed (maybe a few seconds?). 

When I come to think of it maybe the same mechanism should apply for `m_ConnectionInfo` as well to provide a good reflection of the flows currently being handled in TcpReassebmly.

So maybe we can divide the work into 2 phases:

1. Convert `m_ConnectionInfo` into a map
2. Build a cleanup mechanism for closed flows that will cleanup all data structures. We can consider merging `m_ConnectionInfo`, `m_ConnectionList` and `m_ClosedConnectionList` if that makes sense

**Comment by gx740** [comment]
Ok. First I am starting the work on this issue. #250 will be next.

**Comment by seladb** [comment]
makes sense, thank you!

**Comment by gx740** [comment]
From my point of view:
1. The maps `m_ConnectionList`, `m_ClosedConnectionList` and `m_ConnectionInfo` should be synchronized:

- if `m_ConnectionInfo` contains some flow key then only one of `m_ConnectionList` and `m_ClosedConnectionList` may contain the same flow key.
- if `m_ConnectionInfo` does not contain some flow key so the same flow key is not stored in `m_ClosedConnectionList` and `m_ConnectionList`.

2. Cleaning up a list of closed connections can be carry out by `reassemblePacket` method or upon user request. The first way is applied if the user has set a flag indicating whether to remove the connection data after a connection is closed to `true`.

Based on the above items I am suggesting following:

- Add a private member `m_RemoveConnInfoFlag` of type `bool` into `TcpReassembly` class
- Add a private member `m_ClosedConnectionTtl` of type `time_t`. If m_RemoveConnInfoFlag is `true` then the predefined constant is assigned to `m_ClosedConnectionTtl` (for example: 5 seconds). Otherwise the user should put its value when object constructing.
- Add a private member `m_ClosedConnectionTtlList` of type `std::map<time_t, std::list<uint32_t>>` which will contain the flow keys of closed connections. The key of map is a time at which the flow key must be removed from all other maps (`m_ConnectionInfo, m_ClosedConnectionList`).
- Add a public method `void purgeClosedConnections(time_t currentTime)` which will remove expired flow keys.
- `reassemblePacket` method calls `purgeClosedConnections` method once per few seconds. Example of code to be added into the method:
```
if(m_RemoveConnInfoFlag == true)
{
  time_t currentTime = time(NULL);
  if(currentTime  >= m_NextTimeToCheck)
  {
    purgeClosedConnections(currentTime);
    m_NextTimeToCheck = currentTime + Nsecs;
  }
}
```
The reason to use `time()` function instead of `gettimeofday()` is the first one is faster (Refer to: https://stackoverflow.com/questions/6498972/faster-equivalent-of-gettimeofday)

Additional optimizations:
- Remove `m_ClosedConnectionList`
- Do not remove item from `m_ConnectionList` when connection is closed. Just release a memory on which the pointer is point to and assign NULL.
- Change `isConnectionOpen` method as:
```
int TcpReassembly::isConnectionOpen(const ConnectionData& connection) const
{
  std::map<uint32_t, TcpReassemblyData *>::iterator iter = m_ConnectionList.find(connection.flowKey);
  if(iter != m_ConnectionList.end())
    return *iter != NULL;
  return -1;
}
```

Let me know what you think.

**Comment by seladb** [comment]
Sorry for the late response. Overall your proposal looks good to me, here are a few comments:

- instead of adding the `m_RemoveConnInfoFlag` flag, maybe it's better to do the opposite: add a `m_DoNotRemoveConnInfo` flag and by default remove closed connection after a certain period of time
- I'm not sure how you plan to use the map of type `std::map<time_t, std::list<uint32_t>>` to efficiently detect which flows need to be removed at a certain point in time. Could you please elaborate on that?
- Adding `purgeClosedConnections` in `reassemblePacket` might create performance issues. It has some resemblance to a GC (Garbage Collection) scenario where at a certain point in time you might have a lot of flows to cleanup and this might delay the processing of the current packet. I think you should consider a "smart" counter of the maximum connection cleanups you allow in a single iteration of `reassemblePacket`. The reason I say "smart" is because a simple counter might not work in high scale scenarios where the are many connections opening and closing all the time

**Comment by gx740** [comment]
> * instead of adding the `m_RemoveConnInfoFlag` flag, maybe it's better to do the opposite: add a `m_DoNotRemoveConnInfo` flag and by default remove closed connection after a certain period of time

Yes it's better

> * I'm not sure how you plan to use the map of type `std::map<time_t, std::list<uint32_t>>` to efficiently detect which flows need to be removed at a certain point in time. Could you please elaborate on that?

When connection is closed the flow key is inserted into the list of certain item of `m_ClosedConnectionTtl`. Keys are ordered in an ascending order. So we can get a range of values.

Possible implementations of `purgeClosedConnections` method:
```
void purgeClosedConnections(time_t currentTime)
{
  auto iterTimeEnd = m_ClosedConnectionTtlList.upper_bound(currentTime);
  for(auto iterTime = m_ClosedConnectionTtlList.begin(); iterTime != endTimeIter; ++iterTime)
  {
    auto iterKey = iterTime->second.begin(), iterKeyEnd = iterTime->second.end()
    while(iterKey != iterKeyEnd)
    {
      m_ConnectionInfo.erase(*iterKey);
      m_ClosedConnectionList.erase(*iterKey++);
    }
  }

  m_ClosedConnectionTtlList.erase(m_ClosedConnectionTtlList.begin(), iterTimeEnd);
}
```
```
void purgeClosedConnections(time_t currentTime, uint32_t maxNumToRemove = 0)
{
  uint32_t count = 0;
  
  if(maxNumToRemove == 0)
    maxNumToRemove = UINT32_MAX;

  auto iterTimeEnd = m_ClosedConnectionTtlList.upper_bound(currentTime);
  auto iterTime = m_ClosedConnectionTtlList.begin();
  while(iterTime != endTimeIter && count < maxNumToRemove)
  {
    auto &keyList = iterTime->second;
    auto iterKey = keyList.begin(), iterKeyEnd = keyList.end()
    for(; iterKey != iterKeyEnd && count < maxNumToRemove; ++count, ++iterKey)
    {
      m_ConnectionInfo.erase(*iterKey);
      m_ClosedConnectionList.erase(*iterKey);
    }
    
    if(keyList.empty())
      m_ClosedConnectionTtlList.erase(iterTime++);
    else
      ++iterTime;
  }
}

```

Unfortunately we cannot use unordered_map container. It could be significantly faster than map.

**Comment by gx740** [comment]
> * Adding `purgeClosedConnections` in `reassemblePacket` might create performance issues. It has some resemblance to a GC (Garbage Collection) scenario where at a certain point in time you might have a lot of flows to cleanup and this might delay the processing of the current packet. I think you should consider a "smart" counter of the maximum connection cleanups you allow in a single iteration of `reassemblePacket`. The reason I say "smart" is because a simple counter might not work in high scale scenarios where the are many connections opening and closing all the time

You're right. Let's the user will decide when cleanup should be done.

**Comment by gx740** [comment]
I have made changes in class. Code is successfully built under Visual Studio 2019 and g++ but not tested yet.

[TcpReassemblyNew.zip](https://github.com/seladb/PcapPlusPlus/files/3686090/TcpReassemblyNew.zip)

**Comment by gx740** [comment]
An error was fixed in `insertIntoCleanupList`:

[TcpReassemblyNewWithFix.zip](https://github.com/seladb/PcapPlusPlus/files/3686271/TcpReassemblyNewWithFix.zip)

**Comment by seladb** [comment]
Thanks for your answers, here are a few more comments:

- why using a map `typedef std::map<time_t, std::list<uint32_t> > CleanupList` instead of a sorted linked list where each node contains the timestamp where FIN packet was seen and the flow key? If you use a list you can just go over it element by element and delete all flows that are expired up to the current time
- in `purgeClosedConnections()` - letting the user choose how many elements to remove can be confusing because it's difficult to decide on such number in runtime and this number can change over time as traffic increases or decreases. My suggestion would be to either to set the default to a number which is not zero (assuming the user hasn't set the `m_DoNotRemoveConnInfo` flag) and still let a user choose a different number if he/she wants to. You can also decide of this number each time `reassemblePacket` is being called according to the size of the list (although this might be an unnecessary optimization)

**Comment by gx740** [comment]
> * why using a map `typedef std::map<time_t, std::list<uint32_t> > CleanupList` instead of a sorted linked list where each node contains the timestamp where FIN packet was seen and the flow key? If you use a list you can just go over it element by element and delete all flows that are expired up to the current time

In this case the insert operation has a linear complexity because we have to support the list in order. It will be significant impact on performance. When map container is used the complexity is logarithmic (plus one insert into the list with constant complexity). The size of `CleanupList` can be great because the insertion operation is done for each connection that is closed.

**Comment by gx740** [comment]
> * in `purgeClosedConnections()` - letting the user choose how many elements to remove can be confusing because it's difficult to decide on such number in runtime and this number can change over time as traffic increases or decreases. My suggestion would be to either to set the default to a number which is not zero (assuming the user hasn't set the `m_DoNotRemoveConnInfo` flag) and still let a user choose a different number if he/she wants to. You can also decide of this number each time `reassemblePacket` is being called according to the size of the list (although this might be an unnecessary optimization)

I see two possible cases: offline and live traffic.

**Offline traffic**
In this case the number of items are being purged from internal structures does not matter. So the user just call `purgeClosedConnections` without limitation on the number of processed items. `reassemblePacket` can also purge all expired data in certain point of time.

**Live traffic**
In this case the number of items are being purged at time period may have importance. But I think, we cannot choose reasonable amount of elements to be purged because there are parameters with unknown values such as:
- expected traffic intencity
- the depth of input queue(s) of network adapter
- system performance:
  - how many packets per second can be processed by processor core in specific application
  - how many closed connections can be purged per sec

Only the user can choose acceptable threshold.

I suggest to implement two modes of work: automatic purge and manually initiated purge.

The constructor signature is changed as:
```
TcpReassembly(OnTcpMessageReady onMessageReadyCallback,
  void* userCookie = NULL,
  OnTcpConnectionStart onConnectionStartCallback = NULL,
  OnTcpConnectionEnd onConnectionEndCallback = NULL,

  // added parameters
  bool doNotRemoveConnInfo = false,
  uint32_t delayCleanup = 5,
  uint32_t maxNumToClean = 0 // TODO: a constant value can be used instead 0
);
```

**Automatic purge**
- The user sets `doNotRemoveConnInfo` parameter to `false` or uses the default values
- He/she can optionally set other parameters
- `tcpReassembly` calls `purgeClosedConnections` method in only case when `doNotRemoveConnInfo` has false value. Calling is performed once per second.

**Manually initiated purge**
- The user sets `doNotRemoveConnInfo` parameter to `true`
- He/she can optionally set other parameters
- `tcpReassembly` never calls `purgeClosedConnections`
- The user code calls `purgeClosedConnections` and passes values in optional parameters if it's necessary

**Comment by gx740** [comment]
An implementation of ideas described above.

[TcpReassemblyWithAutoCleanup.zip](https://github.com/seladb/PcapPlusPlus/files/3691314/TcpReassemblyWithAutoCleanup.zip)

**Comment by seladb** [comment]
Thanks @gx740 that looks good to me.

Two more comments:

- Instead of having `TcpReassembly::TcpReassembly()` take 7 parameters, maybe it'd better to create a new configuration object (for example `TcpReassemblyConfiguration`) and move your 3 new params to that object
- Please make sure to document these changes well. It's very helpful that the new params have defaults so the average user doesn't have to worry about them, but for the advanced user it should be clear what these params mean

**Comment by seladb** [comment]
@gx740 PR is merged and pushed to `master`, please let me know if we can close this issue

**Comment by gx740** [comment]
You can close this PR.

**Comment by seladb** [comment]
thank you again @gx740 for your continuous contribution to this project, much appreciated!
